src/network/networkd-dhcp4.c

```diff
@@ -142,6 +141,14 @@ static int link_set_dhcp_routes(Link *link) {
                 route->priority = link->network->dhcp_route_metric;
                 route->table = link->network->dhcp_route_table;
 
+                /* Set route scope based on its dst and gw. */
+                if (in_addr_is_localhost(AF_INET, &route->dst) || (address.s_addr && route->dst.in.s_addr == address.s_addr))
+                        route->scope = RT_SCOPE_HOST;
+                else if (in4_addr_is_null(&route->gw.in))
+                        route->scope = RT_SCOPE_LINK;
+                else
+                        route->scope = RT_SCOPE_UNIVERSE;
+
```

Comment:
Done, but I missed your suggestion for the function name... Let me know if you like mine. :P